.github/workflows/firmware-ci.yml

```
@@ -46,9 +56,152 @@ jobs:
+            echo "New dynamic allocation detected in diff; disallowed in hot path." >&2; exit 1; } || true
+
       - name: Generate compile_commands.json
         run: ./agent_runner.sh pio run -t compiledb
 
+      - name: clang-format check
+        run: |
+          find src include -type f \( -name '*.cpp' -o -name '*.hpp' -o -name '*.h' \) -print0 \
+            | xargs -0 clang-format --dry-run --Werror
+
+      - name: clang-tidy check
+        run: |
```

actionlint (1.7.7)

shellcheck reported issue in this script: SC2046:warning:3:52: Quote this to prevent word splitting

(shellcheck)